### PR TITLE
Refactor ut_income_tax formula 

### LINKED ISSUE #4982: Refactor ut_income_tax formula
The current formula for ut_income_tax has some missing calculations. 
Tax form: https://tax.utah.gov/forms/current/tc-40.pdf # line 22 & 32

### PR DESCRIPTION
**Pull request by hua7450**
Fixes #4982

### CODE REVIEW

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py [hunk 1]
```diff
@@ -4,8 +4,14 @@
 class ut_income_tax(Variable):
     value_type = float
     entity = TaxUnit
-    label = "Utah income tax"
+    label = "Utah income tax "
     unit = USD
     definition_period = YEAR
-    adds = ["ut_income_tax_before_refundable_credits"]
+    documentation = "Form TC-40, line 39"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:10]: could you add a reference instead

File: changelog_entry.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+- bump: minor
+  changes:
+    fixed:
+    - Utah income tax formula.
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Refactor the Utah income tax tree.
```

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py [hunk 3]
```diff
@@ -4,8 +4,14 @@
...
     defined_for = StateCode.UT
+
+    def formula(tax_unit, period, parameters):
+        income = tax_unit("ut_income_tax_before_refundable_credits", period)
+        withheld_income_tax = tax_unit("ut_withheld_income_tax", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:15]:
```suggestion
```

Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:15]: dont need it here

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: we should also change:
`ut_subtractions_from_income` to `ut_subtractions`
move `ut_non_refundable_credits` to credits folder in ut

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax_before_non_refundable_credits.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+
+class ut_income_tax_before_non_refundable_credits(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Utah income tax before non-refundable credits"
+    unit = USD
+    definition_period = YEAR
+    documentation = "Form TC-40, line 22"
+    defined_for = StateCode.UT
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax_before_non_refundable_credits.py:11]: and define `ut_income_tax_exempt` for StateCode.UT

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax_before_non_refundable_credits.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    def formula(tax_unit, period, parameters):
+        taxpayer_credit = tax_unit("ut_taxpayer_credit", period)
+        ut_income_tax_before_credits = tax_unit(
+            "ut_income_tax_before_credits", period
+        )
+        is_tax_exempt = tax_unit("ut_income_tax_exempt", period) > 0
+        net_taxpayer_credit = max_(
+            taxpayer_credit - ut_income_tax_before_credits, 0
+        )
+        return where(is_tax_exempt, 0, net_taxpayer_credit)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax_before_non_refundable_credits.py:23]:
```suggestion
        is_tax_liable = ~ tax_unit("ut_income_tax_exempt", period) 
        net_taxpayer_credit = max_(
            taxpayer_credit - ut_income_tax_before_credits, 0
        )
        return is_tax_liable * net_taxpayer_credit
```

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py (hunk 3 above)
Comment (hua7450) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:15]: delete

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py [hunk 6]
```diff
@@ -8,6 +8,7 @@ class ut_taxable_income(Variable):
     unit = USD
     documentation = "Form TC-40, line 9"
     definition_period = YEAR
-    adds = ["ut_total_income"]
     defined_for = StateCode.UT
-    subtracts = ["ut_subtractions_from_income", "ut_state_tax_refund"]
+
+    adds = ["ut_total_income"]
+    subtracts = ["ut_subtractions", "ut_state_tax_refund"]
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py:14]: Should this be capped at 0? I can imagine that even total income - subtractions should be

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py [hunk 7]
```diff
@@ -4,8 +4,15 @@
 class ut_income_tax(Variable):
     value_type = float
     entity = TaxUnit
-    label = "Utah income tax"
+    label = "Utah income tax "
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:7]:
```suggestion
    label = "Utah income tax"
```

File: policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py (hunk 6 above)
Comment (hua7450) [reply policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py:14]: I add the capped 0 condition in `ut_income_tax_before_credits`, as it is represented in the tax form.

**Comment by hua7450** [comment]
The `ut_taxpyaer_credit` (line 20 in [tax from](https://tax.utah.gov/forms/2021/tc-40.pdf)) is not a nonrefundable credit, I remove it from the list. Once I removed it, the result is the same as the TAXISM result.

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py [hunk 8]
```diff
@@ -7,5 +7,12 @@ class ut_income_tax(Variable):
...
-    adds = ["ut_income_tax_before_refundable_credits"]
     defined_for = StateCode.UT
+    reference = (
+        "https://tax.utah.gov/forms/current/tc-40.pdf#page=2"  # line 38
+    )
+
+    def formula(tax_unit, period, parameters):
+        income = tax_unit("ut_income_tax_before_refundable_credits", period)
+        credits = tax_unit("ut_refundable_credits", period)
+        return max_(income - credits, 0)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:18]: This should be uncapped - refundable credits get paid out

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]
run a git pull upstream master

File: policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py (hunk 8 above)
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/ut/tax/income/ut_income_tax.py:18]: use adds/subtracts

File: policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py [hunk 9]
```diff
@@ -8,6 +8,7 @@ class ut_taxable_income(Variable):
     unit = USD
     documentation = "Form TC-40, line 9"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ut/tax/income/taxable_income/ut_taxable_income.py:9]: could you add the tax form reference to all of the variables

**Review by PavelMakarchuk — APPROVED** [review]

**Comment by PavelMakarchuk** [comment]
LGTM after references in vars

**Comment by hua7450** [comment]
In the tax form, the `ut_taxpayer_credit` is computed before other nonrefundable credits, although it is classified as a nonrefundable credit in legal code. However, it is not listed under nonrefundable credits in the instructions. I remove it from our nonrefundable list to maintain consistency with the tax form computation logic.

**Comment by PavelMakarchuk** [comment]
> In the tax form, the `ut_taxpayer_credit` is computed before other nonrefundable credits, although it is classified as a nonrefundable credit in legal code. However, it is not listed under nonrefundable credits in the instructions. I remove it from our nonrefundable list to maintain consistency with the tax form computation logic.

Lets keep the Utah Tax Form structure to align with the Form @MaxGhenis fyi - does not change the logic

**Pull request merged**